analysis/generate_report.py

```diff
@@ -167,7 +170,29 @@ def generate_report(experiment_names,
         experiment_df = data_utils.clobber_experiments_data(
             experiment_df, experiment_names)
 
+    # Generate coverge reports for each benchmark.
+    benchmark_names = experiment_df.benchmark.unique()
     fuzzer_names = experiment_df.fuzzer.unique()
+    coverage_report_directory = os.path.join(report_directory,
+                                             'coverage-reports')
+    if not in_progress:
+        logger.info('Generating coverage reports.')
+        try:
+            if not from_cached_data:
+                report_utils.fetch_source_files(benchmark_names,
+                                                coverage_report_directory)
+                report_utils.fetch_binary_files(benchmark_names,
+                                                coverage_report_directory)
+                report_utils.get_profdata_files(experiment_names[0],
+                                                benchmark_names, fuzzer_names,
+                                                coverage_report_directory)
+
+            # Generate coverage reports for each benchmark.
+            report_utils.generate_cov_reports(benchmark_names, fuzzer_names,
+                                              coverage_report_directory)
+        except Exception:  # pylint: disable=broad-except
+            logger.error('Failed when generating coverage reports.')
```

Comment:
Currently, the input of `generate_report.py` is simply either a database to query or a CSV file (for custom report generation locally). With these change we seem to require binaries, source files, and profdata files as well as input. In order to keep our tools simpler, would it make sense to have a separate tool (eg `generate_clang_coverage_reports.py` for creating the clang html coverage reports? 
- This tool could perhaps take an optional flag pointing to the directory containing those reports to link them (if provided). 
- Alternatively, perhaps it could take exported coverage data in JSON (https://clang.llvm.org/docs/SourceBasedCodeCoverage.html#exporting-coverage-data) output by `generate_clang_coverage_reports.py`? 
- Later, the most ideal solution might be to save the clang coverage data into the DB (i.e., turn the structured JSON data into DB data). That way `generate_report.py` could still simply either take a database to query or a CSV file.

Wdyt?

Reply:
I think it's a good idea. We can also keep the front page simple by moving those differential graphs and reports to another html.
On the other hand, we couldn't avoid using the binaries, source files and profdata files in any case, since we need them to generate the coverage report. The json file can only provide information about covered regions. I generated the json files in the end of measurer and use them to get the differential graphs, this is in another CL.

Reply:
And yes, how long this clang code coverage generation takes, since locally, it might take hours ?

Reply:
I think coverage reports which need binaries can go in their own module.
But I think producing metrics on coverage, like the uniqueness of each fuzzer's coverage should go in generate_report.py

Reply:
> And yes, how long this clang code coverage generation takes, since locally, it might take hours ?

I'm not sure how long it's gonna take to do a coverage generation for a full experiment locally, since I haven't ran a full experiment to get all the source files I need to do this. But I don't think it's going to take hours. I'm using multi-process and even if I don't, it takes approximately 1 sec to generate the report for one pair of fuzzer and benchmark. So I'd say 10 mins or less.

Reply:
> I think coverage reports which need binaries can go in their own module.
> But I think producing metrics on coverage, like the uniqueness of each fuzzer's coverage should go in generate_report.py

Yeah, I agree.

Reply:
So, generate_report.py is a standalone script that anyone can execute locally to generate full report. With this code, that wont be possible, as it relies on measuring files to exist in specific locations of disk. I think this needs to change, we need to put stuff in DB in last cycle ? So that this generate_report.py works quickly as expected. @jonathanmetzman @lszekeres - thoughts ?

Reply:
Putting things in DB would be ideal but I have no idea if the data will be too large. @Shadoom7 it would be nice to have an estimate on how much it would cost to store the regions covered by each trial on the last cycle in the DB (or maybe even fuzzer-benchmark pair to save space).
I totally agree that specific locations on disk is a bad way to do this. I think a reasonable way to do it that wouldn't risk exploding the db is to store info about the GCS bucket in the db (on the experiment entity). Then generate_report can download the data it needs from there.

Reply:
Yes, having the data in the DB would be most ideal, but even before we get there, we could do a simpler to implement alternative mentioned on the top of this thread. Not sure that comment was clear. What i meant was that we could have two separate tools: 1) the existing `generate_report.py`, and 2) a new `generate_clang_coverage_reports.py` that would have this new code this comment thread is on.
- `generate_clang_coverage_reports.py` could output the HTML coverage reports and the JSON coverage data files under some `reports/<experiment>/clang-coverage` directory, provided by a flag.
- `generate_report.py` would take this via an optional --clang-coverage-directory flag, and if provided, then the main report would link to the HTML coverage reports (just like in @Shadoom7's demo) and would create the differential plot.

To answer comments above directly:

> We can also keep the front page simple by moving those differential graphs and reports to another html.

There wouldn't be need for another HTML in this case

> On the other hand, we couldn't avoid using the binaries, source files and profdata files in any case, since we need them to generate the coverage report. 

Yes, the `generate_clang_coverage_reports.py` tool would use these, while `generate_report.py` would use the HTML reports and JSON files output by the former tool.

Having two tools will keep things simpler and will give us more flexibility, even before we put the clang cov data in the db.

Reply:
I think it is still convenient for users to generate report locally, since it syncs all the needed file to a single folder in gcs.
Currently, to generate the report locally. We need to download `gs://my-fuzzbench-reports/<experiment name>/data.csv.gz` to the local `report-dir` and then run generate_report.
With this patch, the only difference is that we need to download `gs://my-fuzzbench-reports/<experiment name>/coverage`, which contains all the source files and binaries.

Reply:
@lszekeres Oh I got you. That makes sense.